app/controllers/forms_controller.rb

```diff
@@ -16,4 +17,24 @@ def update
       render "question_flow/#{step}"
     end
   end
+
+protected
+
+  def check_early_eligibility
+    if gross_income_details_valid?
+      session_data["api_result"] = CfeService.call(session_data, early_eligibility: true)
+      @early_eligibility_result = EarlyEligibility.new(session_data)
+    end
+  end
+
+  def gross_income_details_valid?
```

Comment:
The difficulty with this approach to seeing if we're ready for the early eligibility check (seeing if the relevant forms are valid) is that you need to build in a bunch of knowledge here about which are the relevant forms. It's not just income and other income, it's benefits, and it's potentially all these details for a partner too. You don't want to be re-implementing the logic from the Steps module.

I had thought you could do it by adding an attribute to some Steps::BlahSection classes that say 'when the last step in this section is completed (whatever that is for this particular check - you can work it out dynamically) do/don't do an early eligibility check'. That way in this controller you could simply call 'Steps::Helper.do_the_thinking_for_me(step, session_data)`. But the problem is that Steps::BlahSections don't actually map very well to gross income, disposable income and capital, and _can't_ because the gross income steps don't all follow one another sequentially in the flow, and nor do the disposable income steps.

I wonder if you could update Flow::Handler::STEPS to tag each step with things like `gross_income`, `disposable_income` etc. That way you could see for a given check "have I just completed the last gross income step?" by comparing which steps apply for the current check and what each step is tagged with.

Reply:
Thanks for this! I had a feeling just using the forms in this way wouldn't be quite right - I will look at the Flow::Handler and go from there with your suggestion